src/Umbraco.Core/Scoping/ScopeProvider.cs

```diff
@@ -0,0 +1,352 @@
+            EventsDispatchMode dispatchMode = EventsDispatchMode.Unspecified)
+        {
+            var ambient = AmbientScope;
+            if (ambient == null)
+            {
+                return AmbientScope = new Scope(this, GetNewContext(), isolationLevel, repositoryCacheMode, dispatchMode);
+            }
+
+            // replace noScope with a real one
+            var noScope = ambient as NoScope;
+            if (noScope != null)
+            {
+#if DEBUG_SCOPES
+                Disposed(noScope);
+#endif
+                // peta poco nulls the shared connection after each command unless there's a trx
+                var database = noScope.DatabaseOrNull;
+                if (database != null && database.InTransaction)
+                    throw new Exception("NoScope is in a transaction.");
+                return AmbientScope = new Scope(this, noScope, GetNewContext(), isolationLevel, repositoryCacheMode, dispatchMode);
+            }
+
+            var scope = ambient as Scope;
+            if (scope == null) throw new Exception("Ambient scope is not a Scope instance.");
+
+            return AmbientScope = new Scope(this, scope, isolationLevel, repositoryCacheMode, dispatchMode);
+        }
+
+        private ScopeContext GetNewContext()
+        {
+            return AmbientContext == null
+                ? AmbientContext = new ScopeContext()
+                : null;
+        }
+
+        /// <inheritdoc />
+        public void Reset()
+        {
+            var scope = AmbientScope as Scope;
+            if (scope != null)
+                scope.Reset();
+
+            StaticScopeReference.Dispose();
+        }
+
+        /// <inheritdoc />
+        public ScopeContext Context
+        {
+            get { return AmbientContext; }
+        }
+
+#if DEBUG_SCOPES
+        // this code needs TLC
+        // the idea is to keep in a list all the scopes that have been created
+        // and remove them when they are disposed
+        // so we can track leaks
+
+        // helps identifying when non-httpContext scopes are created by logging the stack trace
+        //private void LogCallContextStack()
+        //{
+        //    var trace = Environment.StackTrace;
+        //    if (trace.IndexOf("ScheduledPublishing") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Scheduled Publishing");
+        //    else if (trace.IndexOf("TouchServerTask") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Server Registration");
+        //    else if (trace.IndexOf("LogScrubber") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Log Scrubber");
+        //    else
+        //        LogHelper.Debug<ScopeProvider>("CallContext: " + Environment.StackTrace);
+        //}
+
+        // helps identifying scope leaks by keeping track of all instances
+        public static readonly List<ScopeInfo> StaticScopeInfos = new List<ScopeInfo>();
```

Comment:
Oh the whole ScopeInfo sure can be greatly optimized, it is as it is because... it is _not_ meant to be ever built unless you want to troubleshoot what's going on, and then perfs don't matter. _Of course_ if we were to ever include it in any kind of build, we'd need to optimize (and maybe debug) it.